Autoload instance analytics content (when EE and avaliable)

escherize opened the pull request:
resolves #27792

I generated this PR by:

1. On a fresh metabase install, set up an "Instance Analytics" collection with a dashboard and a question in it.
2. Manually update its type to be `instance-analytics`
3. Export the `instance-analytics` collection into `resources/instance_analytics`. 
4. With that in place, the mechanism to load it into the app imports the instance analytics content on startup.

Missing behaviors:

- [x] ~Do not run an import if there's unmodified analytics content~ It's OK to re-run
- [x] Figure out how to *create* analytics content.
- [x] ~Fix how login works (if the app sees content it skips the initial signup flow)~ Gonna hold off on this for now
- [x] Tests

Caveats with serialization we should keep in mind:
1. No ability to delete (we might not need this)
2. If a Dashboard has a Card that is not a child of the instance analytics collection, then that dashboard won't be exported. Therefore we must be sure that all the content is actually in the instance analytics collection.

Running this branch with an `instance-analytics` directory in the app resources should yield "Instance Analytics" collection with type `instance-analytics`.

If you can't login after it starts, alter `src/metabase/api/session.clj` like so, and use whatever password you'd like 

```diff
-    (when (u.password/verify-password password (:password_salt user) (:password user))
+    (when true 
```

When you log in, you should have a new audit collection like so:

<img width="705" alt="Screen Shot 2023-06-02 at 4 05 54 PM" src="https://github.com/metabase/metabase/assets/343288/b6658dbb-0c83-4620-a460-a2236be84c92">

escherize committed:
wip

escherize committed:
fix caching behavior for audit db in database-id->connection-pool

escherize committed:
test that caching behavior works as advertised

- db->pooled-connection-spec always returns the same pool for audit-db
- audit-db-id is not in the database-id->connection-pool cache

escherize committed:
DISABLE audit db exposure through GET api/database

escherize committed:
Lots of things

- automagically loads instance analytics on app startup
- TODO: logging in after that is bokred

escherize commented on src/metabase/api/database.clj:
```diff
@@ -284,7 +288,8 @@
                                                       :include-saved-questions-tables? include-saved-questions-tables?
                                                       :include-editable-data-model?    include_editable_data_model
-                                                      :exclude-uneditable-details?     exclude_uneditable_details)
+                                                      :exclude-uneditable-details?     exclude_uneditable_details
+                                                      :include-analytics?              true #_include_analytics)
```
FIXME

escherize committed:
add some logs / fix comment

escherize committed:
Merge branch 'master' into bcm-autoload-instance-analytics

escherize committed:
remove internal_analytics resource

escherize committed:
fix linter issues + update logging

escherize committed:
add tests for existing and missing content

- add test replica of content into test-resources

escherize committed:
remove unused require

escherize committed:
delete sample db + personal collection from test-resources

escherize committed:
get instance_analytics resource off the resource path

escherize committed:
point it to the proper IA resource

escherize committed:
un-hotwire the change to show audit DB on GET /databases

escherize committed:
fix test + lookup the right resource

escherize committed:
remove creator_id references from instance_analytics export test data

- get test feedback

escherize committed:
Revert "remove creator_id references from instance_analytics export test data"

This reverts commit 1aa9d4b1dad304ac0c6d88221f81432d78fc88f0.

escherize committed:
fix tests

escherize committed:
lint

escherize committed:
fix tests

escherize committed:
more test fixing

escherize committed:
linter fix

escherize committed:
Merge branch 'master' into bcm-autoload-instance-analytics

escherize committed:
revert yarn.lock

noahmoss reviewed (commented):
I tried running your branch (with an existing app db) but don't see the Instance Analytics collection. Anything I could be doing wrong?

escherize commented on enterprise/backend/src/metabase_enterprise/serialization/cmd.clj:
```diff
@@ -181,21 +183,43 @@
   (v2.seed-entity-ids/seed-entity-ids!))
+
+
+(comment
```
Yeah, I figured we could leave it here until we move it to a script or part of the release code.

escherize replied:
> I tried running your branch (with an existing app db) but don't see the Instance Analytics collection. Anything I could be doing wrong?

I'll update the PR, but this should only load the instance analytics if there is a directory named `instance_analytics` in the resources. I renamed the test data in test-resources to `instance_analytics_skip` so that it would not load. Then in the tests I point the loader to the `instance_analytics_skip` resource dir to make sure it can load something.

escherize replied:
Done.

escherize committed:
Look up the correct instance_analytics resource

escherize commented on test/metabase/api/collection_test.clj:
```diff
@@ -133,11 +133,13 @@
             (is (= (into public-collections #{"Crowberto Corv's Personal Collection" "Crowberto's Child Collection"})
-                   crowbertos-with-excludes))
+                   ;; TODO: better IA test data
+                   (set (remove #{"Audit" "Instance Analytics"} crowbertos-with-excludes))))
```
I am concerned that there is a race condition, depending on if the test-resource instance analytics ran before this test or not.

I did it this way because if we remove them then it will work either way, but if we always expect them then this will have to run after the instance analytics load.

escherize commented:
> I tried running your branch (with an existing app db) but don't see the Instance Analytics collection. Anything I could be doing wrong?

I'll update the PR, but this should only load the instance analytics if there is a directory named `instance_analytics` in the resources. I renamed the test data in test-resources to `instance_analytics_skip` so that it would not load. Then in the tests I point the loader to the `instance_analytics_skip` resource dir to make sure it can load something.

In fact I noticed an error that I fixed in [c942b60](https://github.com/metabase/metabase/pull/31314/commits/c942b608d4d9ccf8ef938d8e487dc54fcd7e87ef). Nice catch @noahmoss

escherize commented:
So, to see the test data for instance analytics get loaded into the app:

```bash
# checkout this branch (bcm-autoload-instance-analytics)
git checkout -b bcm-autoload-instance-analytics

# copy the test data to a place where it will get picked up
cp -r test_resources/instance_analytics_skip resources/instance_analytics

# start the app as usual
```

There should be a bunch of logs like:
```
2023-06-07 16:55:07,303 INFO v2.load :: Loading Database Audit Database > Schema public > Table report_card > Field enable_embedding
2023-06-07 16:55:07,323 INFO v2.load :: Loading Database Audit Database > Schema public > Table report_card
2023-06-07 16:55:07,326 INFO v2.load :: Loading Database Audit Database
2023-06-07 16:55:07,352 INFO v2.load :: Loading Database Audit Database > Schema public > Table qrtz_simprop_triggers > Field trigger_name
```

escherize committed:
cleanup comment in serialization/cmd.clj

noahmoss reviewed (approved):
Stamping since it now works locally for me. Two remaining questions that we've discussed on Slack:
1. Does the Analytics Database metadata need to be included in the serialized content that gets loaded, or can it just be the new collection & content in it? And similar question for the `settings.yaml` file, not sure why that needs to be there.
2. Can the verbose output when content is loaded on app startup be suppressed?

escherize committed:
silence noisy serialization import logs

escherize committed:
remove a bunch of unused serialization files

escherize committed:
rename Audit Database + docstring

escherize committed:
rename audit database everywhere

escherize committed:
Put the database and tables back

escherize committed:
constrain tests to postgres

escherize committed:
ignore ia data in a test
